File: quantum/languages/java/ChimeraQuantum.java
package chimera.quantum;

public final class ChimeraQuantum {
  public static final class StateVector {
    public final int qubits; public final double[] re; public final double[] im;
    public StateVector(int qubits) { if (qubits < 1) throw new IllegalArgumentException(); this.qubits=qubits; re=new double[1<<qubits]; im=new double[1<<qubits]; re[0]=1; }
    public double probabilitySum(){ double s=0; for(int i=0;i<re.length;i++) s+=re[i]*re[i]+im[i]*im[i]; return s; }
    public void hadamard(int target){ if(target<0||target>=qubits) throw new IndexOutOfBoundsException(); int bit=1<<target; double k=1/Math.sqrt(2); for(int i=0;i<re.length;i++) if((i&bit)==0){int j=i|bit; double xr=re[i],xi=im[i],yr=re[j],yi=im[j]; re[i]=k*(xr+yr); im[i]=k*(xi+yi); re[j]=k*(xr-yr); im[j]=k*(xi-yi);} }
  }
}

File: quantum/languages/kotlin/ChimeraQuantum.kt
package chimera.quantum

object ChimeraQuantum {
  class StateVector(val qubits: Int) {
    val re = DoubleArray(1 shl qubits)
    val im = DoubleArray(1 shl qubits)
    init { require(qubits > 0); re[0] = 1.0 }
    fun probabilitySum() = re.indices.sumOf { re[it]*re[it] + im[it]*im[it] }
    fun hadamard(target: Int) {
      require(target in 0 until qubits); val bit=1 shl target; val k=1.0/kotlin.math.sqrt(2.0)
      for(i in re.indices) if(i and bit == 0){ val j=i or bit; val xr=re[i]; val xi=im[i]; val yr=re[j]; val yi=im[j]; re[i]=k*(xr+yr); im[i]=k*(xi+yi); re[j]=k*(xr-yr); im[j]=k*(xi-yi) }
    }
  }
}
